Plugin/LSP/LSPOutlineView.h

```diff
@@ -0,0 +1,199 @@
+#define kConfigOutlineStartSubclassesCollapsed  "OutlineStartSubclassesCollapsed"
+#define kConfigOutlineStartFunctionsCollapsed   "OutlineStartFunctionsCollapsed"
+#define kConfigOutlineStartVariablesCollapsed   "OutlineStartVariablesCollapsed"
+#define kConfigOutlineStartContainersCollapsed  "OutlineStartFunctionsCollapsed"
+#define kConfigOutlineKeepNamespacesExpanded    "OutlineKeepNamespacesExpanded"
+#define kConfigOutlineSortType                  "OutlineSortType"
+#define kConfigOutlineShowSymbolDetails         "OutlineShowSymbolDetails"
+#define kConfigOutlineShowSymbolKinds           "OutlineShowSymbolKinds"
+#define kConfigOutlineForceTree                 "OutlineForceTree"
+
+class WXDLLIMPEXP_SDK LSPOutlineView : public wxPanel
+{
+public:
+    enum class MODE {
+        DOCUMENT_SYMBOL,
+        SYMBOL_INFORMATION,
+        SYMBOL_INFORMATION_PARSED,
+    };
+    enum Style {
+        STYLE_DEFAULT               = 0,
+        STYLE_NO_SEARCHBAR          = 1 << 0,
+        STYLE_SMALL_FONT            = 1 << 1,
+        STYLE_ACTIVATE_ON_SELECTION = 1 << 2
+    };
+    LSPOutlineView(wxWindow* parent, wxWindowID id = wxID_ANY, long style = LSPOutlineView::STYLE_DEFAULT);
+    virtual ~LSPOutlineView();
+
+    void SetSymbols(const std::vector<SymbolInformation>& symbols, const wxString& filename);
+    void SetSymbols(const std::vector<DocumentSymbol>& symbols, const wxString& filename);
+    void SetEmptySymbols();
+    void SetEmptyMessage(const wxString& message) { m_emptyMessage = message; }
+    void SetLoadingMessage(const wxString& message) { m_loadingMessage = message; }
+    /**
+     * @brief Sorts the current DocumentSymbol. Does not refresh the tree - use DoInitialiseDocumentSymbol after!
+     * @param sort
+     */
+    void SortSymbols(LSPSymbolParser::SortType sort);
+    void CollapseTree();
+    void ExpandTree();
+    
+protected:
+    wxString m_currentSymbolsFileName;
+    std::vector<DocumentSymbol> m_documentSymbols;
+    std::vector<SymbolInformation> m_symbolsInformation;
+    enum ID {
+        ID_SHOW_DETAILS,
+        ID_SHOW_KIND,
+        ID_SORT_KIND_NAME,
+        ID_SORT_KIND_NAME_CASE_SENSIIVE,
+        ID_SORT_LINE,
+        ID_TOOL_SORT,
+        ID_TOOL_DETAILS,
+        ID_TOOL_KIND,
+        ID_TOOL_COLLAPSE_ALL,
+        ID_TOOL_EXPAND_ALL,
+        ID_TOOL_MENU,
+        ID_MENU_KEEP_NAMESPACE_EXPANDED,
+        ID_MENU_FORCE_TREE,
+        ID_MENU_INITIALISE_ALL_COLLAPSED,
+        ID_MENU_INITIALISE_ENUMS_COLLAPSED,
+        ID_MENU_INITIALISE_FUNCTIONS_COLLAPSED,
+        ID_MENU_INITIALISE_CLASSES_COLLAPSED,
+        ID_MENU_INITIALISE_SUBCLASSES_COLLAPSED,
+        ID_MENU_INITIALISE_CONTAINERS_COLLAPSED,
+        ID_MENU_INITIALISE_VARIABLES_COLLAPSED
+    };
+    struct TreeState {        
+        // Pointer to DocumentSymbol wrapped in wxTreeItemId
+        wxTreeItemId selectedNode;
+        bool manuallyCollapsed = false;
```

Comment:
Here and everywhere, prefer `{..}` initialization, e.g. `bool manuallyCollapsed{false};`